contracts/finance/ProtocolStaking.sol

```diff
@@ -0,0 +1,263 @@
+        return _unstakeRequests[account].latest() - _totalReleased[account];
+    }
+
+    /// @notice Returns the recipient for rewards earned by `account`.
+    function rewardsRecipient(address account) public view virtual returns (address) {
+        address storedRewardsRecipient = _rewardsRecipient[account];
+        return storedRewardsRecipient == address(0) ? account : storedRewardsRecipient;
+    }
+
+    function _stake(uint256 amount) internal virtual {
+        _updateRewards();
+        _updateRewards(msg.sender);
+
+        if (isOperator(msg.sender)) {
+            uint256 previousStakedAmount = balanceOf(msg.sender);
+            uint256 newStakedAmount = previousStakedAmount + amount;
+
+            _totalStakedWeight = _totalStakedWeight + weight(newStakedAmount) - weight(previousStakedAmount);
+        }
+
+        _mint(msg.sender, amount);
+        IERC20(stakingToken()).safeTransferFrom(msg.sender, address(this), amount);
+
+        emit TokensStaked(msg.sender, amount);
+    }
+
+    function _unstake(uint256 amount) internal virtual {
+        _updateRewards();
+        _updateRewards(msg.sender);
+
+        require(amount != 0, InvalidAmount());
+
+        if (isOperator(msg.sender)) {
+            uint256 previousStakedAmount = balanceOf(msg.sender);
+            uint256 newStakedAmount = previousStakedAmount - amount;
+
+            _totalStakedWeight = _totalStakedWeight + weight(newStakedAmount) - weight(previousStakedAmount);
+        }
+
+        _burn(msg.sender, amount);
+
+        if (_unstakeCooldownPeriod == 0) {
+            IERC20(stakingToken()).safeTransfer(msg.sender, amount);
+        } else {
+            (, uint256 lastReleaseTime, uint256 totalRequestedToWithdraw) = _unstakeRequests[msg.sender]
+                .latestCheckpoint();
+            uint256 releaseTime = block.timestamp + _unstakeCooldownPeriod;
+            _unstakeRequests[msg.sender].push(
+                uint48(Math.max(releaseTime, lastReleaseTime)),
+                uint208(totalRequestedToWithdraw + amount)
+            );
+        }
+
+        emit TokensUnstaked(msg.sender, amount);
+    }
+
+    function _updateRewards(address account) internal virtual {
+        if (_userStakingInfo[account].rewardsPerUnitPaid == 0) return;
+        _userStakingInfo[account] = UserStakingInfo({rewards: earned(account), rewardsPerUnitPaid: _rewardsPerUnit});
+    }
+
+    function _updateRewards() internal virtual {
+        if (block.timestamp == _lastUpdateTimestamp) {
+            return;
+        }
+
+        uint256 secondsElapsed = block.timestamp - _lastUpdateTimestamp;
+        _lastUpdateTimestamp = block.timestamp;
+
+        if (_totalStakedWeight == 0) {
+            return;
+        }
+
+        uint256 rewardsPerUnitDiff = (secondsElapsed * _rewardRate * 1e18) / _totalStakedWeight;
+        _rewardsPerUnit += rewardsPerUnitDiff;
+        _lastUpdateTimestamp = block.timestamp;
+    }
```

Comment:
Do we really want this ? This function means that for every block where an operation happens, we will ahve to do expensive sstores. Is it not possible to compute the value using view function and only update the stored value when absolutelly necessary (like when _rewaredRate changes)